### PR TITLE
Add aura/di basics

### PR DESCRIPTION
**Pull request by harikt**

### CODE REVIEW

**Comment by harikt** [comment]
Fixes https://github.com/zendframework/zend-expressive/issues/109

File: doc/book/container/bookdown.json [hunk 1]
```diff
@@ -5,6 +5,6 @@
         {"Container Factories": "factories.md"},
         {"Using zend-servicemanager": "zend-servicemanager.md"},
         {"Using Pimple": "pimple.md"}
```
Comment (samsonasik) [inline doc/book/container/bookdown.json]: missing `,` before next line

File: doc/book/container/bookdown.json [hunk 2]
```diff
@@ -5,6 +5,6 @@
...
+        {"Using Aura.Di": "aura-di.md"}
     ]
 }
-
```
Comment (samsonasik) [inline doc/book/container/bookdown.json]: new line of eof ending

File: doc/book/container/aura-di.md [hunk 3]
```diff
@@ -0,0 +1,119 @@
...
+- lazy-loaded instances, services, includes/requires, and values
+
+- instance factories
+
+- optional auto-resolution of typehinted constructor parameter values
+
+
+## Installing Aura.Di
+
+Aura.Di v3 only implement [container-interop](https://github.com/container-interop/container-interop).
```
Comment (samsonasik) [inline doc/book/container/aura-di.md]: /s/implement/implements

File: doc/book/container/bookdown.json (hunk 1 above)
Comment (harikt) [inline doc/book/container/bookdown.json]: good catch.

File: doc/book/container/bookdown.json (hunk 2 above)
Comment (harikt) [inline doc/book/container/bookdown.json]: Hm... new line needed?

File: doc/book/container/bookdown.json [hunk 4]
```diff
@@ -4,7 +4,7 @@
         {"Introduction": "intro.md"},
         {"Container Factories": "factories.md"},
         {"Using zend-servicemanager": "zend-servicemanager.md"},
-        {"Using Pimple": "pimple.md"}
+        {"Using Pimple": "pimple.md"},
+        {"Using Aura.Di": "aura-di.md"}
     ]
 }
-
```
Comment (samsonasik) [inline doc/book/container/bookdown.json]: yup, new line for end of file, a blank line after `}`

**Comment by weierophinney** [comment]
My principal suggestion is that you follow the same patterns used for the ZF2 service manager and Pimple: in other words, have a `config/services.php` file that returns the fully configured container. We plan to use this pattern in the skeleton application, with example files for each container we document; this allows keeping the bootstrap file the same across implementations.

**Comment by harikt** [comment]
@weierophinney question : 

Do you think changing to `config/services.php` 

```php
use Aura\Di\ContainerBuilder;

$container_builder = new ContainerBuilder();

// use the builder to create and configure a container
// using an array of ContainerConfig classes
$di = $container_builder->newConfiguredInstance([
    'Application\_Config\Common',
]);
return $di;
```

and having a seprate class is acceptable? The things are you can load other config classes if they are already provided by the package and don't need to add the config our own.

```
<?php
chdir(dirname(__DIR__));
$container = require 'config/services.php';
$app = $container->get('Zend\Expressive\Application');
$app->run();
```

If this is acceptable let me know.

**Comment by weierophinney** [comment]
Absolutely; that's exactly how we handled the other two, and why I
suggested it. :-)

**Comment by harikt** [comment]
Have made the changes @weierophinney . Hope you like. Feel free to make the necessary changes.

Thank you.

File: doc/book/container/aura-di.md [hunk 5]
```diff
@@ -0,0 +1,118 @@
...
+// using an array of ContainerConfig classes
+// make sure the classes can be autoloaded
+return $container_builder->newConfiguredInstance([
+    'Application\_Config\Common',
+]);
+```
+
+The bare minimal `ContainerConfig ` code needed to make zend-expressive work is
+
+```php
```
Comment (weierophinney) [inline doc/book/container/aura-di.md]: Where should this class file be placed?

File: doc/book/container/bookdown.json (hunk 4 above)
Comment (weierophinney) [inline doc/book/container/bookdown.json]: @samsonasik No blank line at the end of the file, please!!!!

File: doc/book/container/aura-di.md (hunk 5 above)
Comment (weierophinney) [inline doc/book/container/aura-di.md]: I'm assuming `src/_Config/Common.php`?

Comment (harikt) [inline doc/book/container/aura-di.md]: Normally in `config` , and add autoload to composer.json . See skelton https://github.com/harikt/expressive-aura-skelton/blob/bbbbf233170ee4b76877860d676116f44f1d75d2/composer.json#L11 .

**Pull request merged**